Address Bill's review list from April 04

wmamills opened the pull request:
This PR attempts to address items in my review list on April 04.
https://lists.linaro.org/archives/list/virtio-msg@lists.linaro.org/thread/LRPZDHB44SRIYIVEDUBWXME35LNF5BEM/

PDF rendered version of this PR: (Updated May 9)
https://github.com/wmamills/virtio-msg-spec/releases/download/virio-msg-draft-v10/virtio-v1.4-wd01.pdf

Currently this PR addresses:
* Eliminate terms Frontend and Backend
* Make it clear that EVENT_ messages don't have a response
* Replace Device ID with Device Number
* Remove Bus Level Reset and STATUS
* Add clarifications for maximum message size

Added April 24:
* add section on bus specific messages, mention OoB notifications and shared memory setup
* make EVENT_ messages mandatory for common transport to bus implementation POV
  (bus has choice of send inband, convert to OoB, or drop/poll)

On April 27 (pushed April 28 AM):
* fixed spelling issues and added missed changes for all the topics in above
* handle device reset according to spec
* add number of virtqueues to GET_DEVICE_INFO
* clarify feature handling as discussed in April 24 meeting
* return new device status in GET_DEVICE_STATUS
* change wording used for empty payload responses
* In GET_VQUEUE, clarify that 0 maximum size mean not usable
* SET_VQUEUE does not echo back the result as it cannot change
* Changing features after init is not allowed
* Add info about changing virtqueues after init is done
* Note that a device status change w/o config change also causes EVENT_CONFIG
* SET_CONFIG details for how to reject gen count mismatch and note that resulting data may not match set data
* More notes about how config gen count works

On April 28 PM:
* fix-up wording for prior commits for feature handling, no feature renegotiation, SET_CONFIG, and max message size
* Move EVENT_ to 0x40
* Remove error msg & reserved ID 1
* add EVENT_CONFIG0 and define how to handle config changes that cant fit in one EVENT_CONFIG
* don't assume config gen count starts at 0
* add clarity for device num bit packing in GET_DEVICES
* add virtqueue index == -1 means all virtqueues
* EVENT_AVAIL extra fields are 0 if not negotiated
* EVENT_AVAIL next wrap make single bit like other transports
* fix-up the conformance section for notifications being required from transport level POV

On April 29 AM:
* fixed typo that prevented pdf generation
* Added link to pdf in this message

Things to do in a different PR:
* specify rules about don't look beyond message length (or assume zero)
* Lower max message size?
* what should we use for MUST etc, \emph{MUST} \textbf{MUST} or just MUST
  (looks like existing spec just uses MUST)
* DEVICE_REMOVED/ADDED to BUS_MSG_EVENT_DEVICE(added, removed)
* add SHM support
* add flags to virtqueue info for admin queue support

On May 9: Address most of the review comments:
* global search & replace GET_DEVICE_FEATURES/SET_DEVICE_FEATURES
* add "unique" to device number
* remove "likely" from description of set config race & fix typos
* remove extra reference to bus message section
* remove "implementation" from driver side bus
* mention feature bit that enables extra fields in EVENT_AVAIL
* replace cut & paste "driver" with correct "device"
* mention EVENT_CONFIG0 as other option for notification
* simplify wording around device re-init
* reword SET_CONFIG(0)
* use "event messages" instead of "messages that have no response"
  also use "request/response messages" for the other case
* mention that GET_DEVICE_INFO is the only valid message for an inactive device and specify that it should return all zeros
* remove redundant procedures in SET_CONFIG message definition and replace with reference
* fix spelling of "preferred" and "notify"
* repeat "and the device" to add clarify to EVENT_CONFIG0 def
* rename EVENT_CONFIG0 to EVENT_CONFIG_EMPTY

wmamills committed:
virtio-msg: Eliminate the terms frontend and backend.

The virtio spec does not use the terms frontend and backend and
neither should we.  Replace frontend with driver and backend with device.

Frontend is easy to replace with driver and should be understood by all.

The term backend is not defined or used the same in industry.
Normally it means the whole device side implementation including
the virtual device, the bridge to a real device, and the driver for the
real device.  Here we are only interested in the virtual device
implementation.

At the spec level the terms "driver" and "device" are collective term for
all the software on that side and would include the device type code,
the common virtio-msg transport level and the virtio-msg bus specific level.

If we need to specifically call out an action from the bus level, use the
words "driver side bus" and "device side bus" respectively.

Signed-off-by: Bill Mills <bill.mills@linaro.org>

wmamills committed:
virtio-msg: Make it clear that no response is possible for EVENT_ etc.

Replace wording like "does not require a response" with.
"there is no response".

The first wording makes it sound like a response is possible but optional.
The second wording makes it sound like no response is possible which is
what we want.

The sender needs to know ahead of time if a response is expected.  So making
it optional has not purpose.  Eliminate variation and make it clear.

Signed-off-by: Bill Mills <bill.mills@linaro.org>

wmamills committed:
virtio-msg: Replace Device ID with Device Number

Device ID is already used in the virtio spec as a synonym for device type.
Change our device instance identifier to "device number".

Rename the message header field to dev_num.

Signed-off-by: Bill Mills <bill.mills@linaro.org>

wmamills committed:
virtio-msg: Remove bus level reset and STATUS for now.

It was decided at the April 10th meeting that this requires more discussion
as we need to handle graceful shutdown and unexpected reset.  We decided
to focus on the device level now and we should remove this bus level status
and reset for now.

Signed-off-by: Bill Mills <bill.mills@linaro.org>

wmamills committed:
virtio-msg: add clarifications for maximum message size

* Maximum message size is from the POV of the common transport and does
  not include any bus specific overhead
* message sizes > 274 are not expected to be useful

Signed-off-by: Bill Mills <bill.mills@linaro.org>

wmamills committed:
virtio-msg: add section on bus specific messages.

* Describe bus specific messages.
* Mention possibility to
  * config Out-of-band notifications
  * setup shared memory regions
  * error handling
  * debug and error logging

Signed-off-by: Bill Mills <bill.mills@linaro.org>

wmamills committed:
virtio-msg: notifications are mandatory from common transport POV

Change the language to show that the notification messages always exist
between the common transport layer and the bus implementation.

Show that the bus implementation has the choice of sending the event
messages in band, converting them to OoB events, or (last resort)
dropping them and reinventing them on the other side via pooling.

Signed-off-by: Bill Mills <bill.mills@linaro.org>

wmamills committed:
virtio-msg: handle device reset according to existing spec

Device initiated reset is done by setting the DEVICE_NEEDS_RESET bit in
the device status.  The driver then takes action.

Also make it clean that the driver only need to re-init the device if it
wants to use it again.  It is valid to leave it in the reset state if not.

Signed-off-by: Bill Mills <bill.mills@linaro.org>

wmamills committed:
virtio-msg: add number of virtqueues to GET_DEVICE_INFO

Add the number of virtqueues to device info.  This allows the driver
to skip querying virtqueues that do not exist.

Signed-off-by: Bill Mills <bill.mills@linaro.org>

wmamills committed:
virtio-msg: clarify feature handling

* Change the message names to GET_DEVICE_FEATURES and SET_DRIVER_FEATURES
  to make it clear that what you set and what you get are different
* No need to echo back the data on SET_DRIVER_FEATURES and there is no
  per bit NAK
* Add a note about what a device should do if it does not like the features
  This is as already defined in the standard
* Make it clear that features come in blocks of 32 bits

Q: In GET_DEVICE_INFO, should it return number of feature _bits_ as it does
   now or return number of feature _blocks_?

Signed-off-by: Bill Mills <bill.mills@linaro.org>

wmamills committed:
virtio-msg: SET_DEVICE_STATUS returns new status

Return the new status as it may be different than the one set.

Signed-off-by: Bill Mills <bill.mills@linaro.org>

wmamills committed:
virtio-msg: clarify that an empty response is still a response

"Answer 0 0 None" may cause some to question if a response is needed.
The text says that a response is needed but change the wording here to
make it clear that the response is required but the payload is empty.

Signed-off-by: Bill Mills <bill.mills@linaro.org>

wmamills committed:
virtio-msg: Clarify GET_VQUEUE

Clarify that a virtqueue that has zero maximum size does not really exist
and must not be used.

Signed-off-by: Bill Mills <bill.mills@linaro.org>

wmamills committed:
virtio-msg: SET_VQUEUE should not echo the data back

SET_VQUEUE does not need to echo the data set as it cannot change.
Drop the echoed data.

Signed-off-by: Bill Mills <bill.mills@linaro.org>

wmamills committed:
virtio-msg: Changing features after init time are not allowed

* Make the ConfigUpdates section about config changes only
* Describe device side config changes also

Signed-off-by: Bill Mills <bill.mills@linaro.org>

wmamills committed:
virtio-msg: add info about virtqueue changes at runtime

* Add section about virtqueue changes after initial init
* Add info about VIRTIO_F_RING_RESET feature

Signed-off-by: Bill Mills <bill.mills@linaro.org>

wmamills committed:
virtio-msg: A device status change is a cause for EVENT_CONFIG

The EVENT_CONFIG message can be sent when the device changes the device
status without any config change.

Signed-off-by: Bill Mills <bill.mills@linaro.org>

wmamills committed:
virtio-msg: SET_CONFIG gen count rejection details

* Define a way for the response SET_CONFIG response to declare the set did
  not work.  (This replaces the previous use of the ERROR message.
* Add more detail about the SET_CONFIG retry and how it overlaps with
  EVENT_CONFIG
* Also make it clear that some data may not match the data written even
  though the SET_CONFIG was not rejected

Signed-off-by: Bill Mills <bill.mills@linaro.org>

wmamills committed:
virtio-msg: add clarity about how the generation count works

Make it clear that the count does not increment for each byte changed
but instead increments at least once for each driver visible change set.

Signed-off-by: Bill Mills <bill.mills@linaro.org>

wmamills committed:
virtio-msg: fix spelling of reserved

This existing typo was found while fixing many of my own.

Signed-off-by: Bill Mills <bill.mills@linaro.org>

wmamills committed:
virtio-msg: move no response messages IDs to be >= 0x40

Messages that don't have a response should have bit 6 set (ID & 0x40) != 0

Change the IDs and explain the convention.

Signed-off-by: Bill Mills <bill.mills@linaro.org>

wmamills committed:
virtio-msg: Remove ERROR message from transport and bus std

Remove the ERROR message from the trasport messages.
Remove the ERROR message from standardized bus messages.
Reserve ID 1 in the bus for standardized error ID in the future.
Reserve ID 1 at transport in case we have a similar global message need.

Signed-off-by: Bill Mills <bill.mills@linaro.org>

wmamills committed:
virtio-msg: add EVENT_CONFIG0

Add this message for use by driver bus implementations that do polling or
have a single bit notification for status and config changes.

Also use it if an atomic config update can not fit in one EVENT_CONFIG
message.  (However in most cases this can be handled by breaking the
change into multiple EVENT_CONFIG messages will unique gen counts.

Signed-off-by: Bill Mills <bill.mills@linaro.org>

wmamills committed:
virtio-msg: don't assume configuration generation count == 0 at start

The driver side should not assume the config gen count starts are zero.
The device should not reset the count to zero at device reset.

Signed-off-by: Bill Mills <bill.mills@linaro.org>

wmamills committed:
virtio-msg: add clarity to the bit packing for GET_DEVICES

* Make it clear that the offsets and numbers should be multiples of 8.
* Make it more clear how the bits are packed into the byte sequence.

Signed-off-by: Bill Mills <bill.mills@linaro.org>

wmamills committed:
virtio-msg: Add virtqueue index = -1 meaning

* Add virtqueue index == -1 means all virtques

Signed-off-by: Bill Mills <bill.mills@linaro.org>

wmamills committed:
virtio-msg: EVENT_AVAIL next offset & wrap clarifications

next offset and wrap are only available when VIRTIO_F_NOTIFICATION_DATA
has been negotiated.  This data will not be available when using OoB
notifications so the bus should prevent this feature from being negotiated.

Signed-off-by: Bill Mills <bill.mills@linaro.org>

wmamills committed:
virtio-msg: EVENT_AVAIL make "next wrap" a single bit like other transports

In other transports the "next wrap" field is a single bit and mapped to the
MSB of a 16 bit value.  We don't need a whole 32 bit value just for the
wrap value and having it this way causes confusion with how to set it.

Although we have a 32 bit value do the same as other transports and map the
wrap bit to the MSB and use the rest as the index.

Signed-off-by: Bill Mills <bill.mills@linaro.org>

wmamills committed:
virtio-msg: additional cleanup for notifications being required

Clean-up the conformance section for the notification messages being
requires from the transport level POV.

Signed-off-by: Bill Mills <bill.mills@linaro.org>

vireshk commented on transport-msg.tex:
```diff
@@ -982,21 +1036,31 @@ \subsubsection{Overview}
 
```
Maybe we can keep the order of the fields same as GET_CONFIG ? i.e. offset, size, generation, data.

wmamills replied:
OK but I would vote we change GET_CONFIG to match this (gen, offset, size, data)
gen is always good info but often offset & size are 0 and data is empty

arnopo commented on transport-msg.tex:
```diff
@@ -1056,8 +1056,8 @@ \subsection{Bus Messages}\label{sec:Virtio Transport Options / Virtio Over Messa
 
-This message is sent by the virtio-msg bus frontend and requires a response
-from the backend.
+This message is sent by the virtio-msg driver side bus and requires a response
```
I wonder if 'driver/device side bus' might be confusing to disociate the transport from the bus. From my perspective, 'driver/device transport bus' or 'driver/device virtio-msg transport bus' could be more explicit.

Anyway, it is probably a point that will trigger discussions on the Oasis PR, so keeping it like this for the moment is okay with me

arnopo commented on transport-msg.tex:
```diff
@@ -612,7 +612,7 @@ \subsubsection{Driver Notifications}
         (depending on the negotiated features), so the device can jump directly
         to the updated buffers.
-  \item \textbf{Response}: The \msgref{EVENT_AVAIL} message does not require a
+  \item \textbf{Response}: The \msgref{EVENT_AVAIL} message does not have a
```
`SHALL NOT receive a response`? 
if ok with that there are several instances

bertrand-marquis replied:
The bus might require a response anyway (we have this case in FF-A with direct messages where an event might get an answer anyway as direct messages always include a response) hence the original "does not require" to leave the bus handle it. Now this is a bus level issue so shall not sounds good to me.

wmamills replied:
I would say that the FF-A bus should discard any response to those messages (and declare/log an error if they are not empty).  From the transport level they do not exist.

bertrand-marquis replied:
agree that it is bus specific and the response never comes back to the transport

wmamills replied:
I am leaving this as-is

arnopo commented on transport-msg.tex:
```diff
@@ -1064,10 +1065,10 @@ \subsection{Bus Messages}\label{sec:Virtio Transport Options / Virtio Over Messa
 \hline \hline
 Request & 0 & 2 & Offset \\
-        & 2 & 2 & Number of device IDs requested \\
+        & 2 & 2 & Number of device numbers requested \\
```
This sentence is quite strange, and with the description above, it gives me the impression that 'device number' is too vague

In addition the term "device number is only used in the spec"
```
5.5 Traditional Memory Balloon Device
This is the traditional balloon device. **The device number 13** is reserved for a new memory balloon interface,
with different semantics, which is expected in a future version of the standard
```
```
B.3 What Device Number?
**Device numbers** can be reserved by the OASIS committee: email virtio-comment@lists.linux.dev (after
subscribing through virtio-comment+subscribe@lists.linux.dev) to secure a unique one.
Meanwhile for experimental drivers, use 65535 and work backwards.
```

Regarding PCI (`4.1.2 PCI Device Discovery`) there is a notion of `PCI device ID`. Could we use same approach usinf "MSG device id" or "MSG device reference"?

arnopo commented on transport-msg.tex:
```diff
@@ -598,12 +598,16 @@ \subsection{Device Operation}
-it (e.g., if a feature akin to \emph{NOTIFY_ON_AVAIL} is defined or if
-\emph{NEGOCIATION\_DATA} was agreed upon).
+defines the message, \msgref{EVENT_AVAIL}, that the driver will use to
+notify the device of pending buffers.
```
The `EVENT _AVAIL` is not used by the driver but by the driver side bus
suggestion:
```
that the driver bside bus \enph{MAY}l use to notify the device of pending buffers.
```

wmamills replied:
This is not correct.  The driver uses the transport specific virtqueue kick op and for virtio-msg transport that results in EVENT_AVAIL message.  The driver side bus may send EVENT_AVAIL in band or convert it to an out-of-band notification, (or discard it but that is sub-par).

arnopo replied:
Right, it is the transport not the bus

come back to the full sentence

```
After setting up one or more virtqueues, the driver \emph{MUST} send notifications
to signal that new buffers are available for processing. The virtio-msg transport
defines the message, \msgref{EVENT_AVAIL}, that the driver will use to
notify the device of pending buffers.

```
The remaining point is the use of the terms MUST and "will" regarding
 `2.7.10.1 Driver Requirements: Available Buffer Notification Suppression`.  

We could replace by similar sentence that `2.7.13 Supplying Buffers to The Device`

```
After setting up one or more virtqueues, the driver \emph{MAY} send an available buffer notification to the device if such notifications are not suppressed. The virtio-msg transport defines the message, \msgref{EVENT_AVAIL}, to notify the device of available buffers.
```

arnopo commented on transport-msg.tex:
```diff
@@ -598,12 +598,16 @@ \subsection{Device Operation}
+out-of-band (OoB) notification.  If not using OoB notifications, the driver
+side bus \emph{SHOULD} send the messages via the normal message channel.
+As a final option, the bus \emph{MAY} discard the messages but only if it knows
+that the device will poll the virtqueues directly.
```
Alternatively, the driver side bus ....

wmamills replied:
I think you are saying drop the word "implementation" and I agree it is not needed.

arnopo commented on transport-msg.tex:
```diff
@@ -617,15 +621,24 @@ \subsubsection{Driver Notifications}
 
-During normal operation, the device \emph{MAY} send notifications to the driver
-about configuration changes or the completion of buffers in virtqueues.
+During normal operation, the driver side bus \emph{MUST} send notifications to
```
```suggestion
During normal operation, the device side bus \emph{MUST} send notifications to
```

bertrand-marquis replied:
+1

arnopo commented on transport-msg.tex:
```diff
@@ -617,15 +621,24 @@ \subsubsection{Driver Notifications}
+    \item Manufactured by the driver side bus periodically (polling).
+        This option should be reserved only for situation where nothing else
+        will work.
+  \end{itemize}
```
This enumeration is not clear to me.

bertrand-marquis replied:
The last manufactured by the driver periodically I do not get:
- do you mean the bus is generating "fake" events to force the drivers to poll ?
In FF-A we have kind of this case possible and the bus is polling periodically the device side for pending event so that it can generate events to the transport.
I do see how periodical fake events could work for virtqueues but not really for config for example.
Maybe something to discuss.

wmamills replied:
Yes the bus invents messages.
"but not really for config for example."  
Yes it works for device status and config changes.  That is what EVENT_CONFIG0 was added for.  There are other was to do it like a bus message that is ASYNC_POLL

wmamills replied:
I set this as a topic for the sprint.
I am happy to reword to make my intention clear or we can decide polling should be done differently

arnopo commented on transport-msg.tex:
```diff
@@ -683,16 +683,17 @@ \subsubsection{Device Reset and Shutdown}
+\emph{MAY} reinitialize the device (see
 \ref{sec:Virtio Transport Options / Virtio Over Messages / Device Initialization})
-before resuming normal operation.
+if it wishes to try and use the device again.
```
not sure if the sentence is correct.
`if it wishes to try to use the device again.`?

wmamills replied:
it is awkward.  I will try to do better

wmamills replied:
Replaced: "if it wishes to try to use the device again"
With: "if it wishes to use the device again"
in two places.

I am open to further wording improvements if anyone has them

arnopo commented on transport-msg.tex:
```diff
@@ -683,16 +683,17 @@ \subsubsection{Device Reset and Shutdown}
+an \msgref{EVENT_CONFIG} with a \emph{device status} with the DEVICE_NEEDS_RESET
+bit set.  The driver \emph{SHOULD} reset the device after this using a
+\msgref{SET_DEVICE_STATUS}(0) message.  The driver \emph{MAY} reinitialize the
+device if it wishes to try and use the device again.
```
same here

wmamills replied:
see above

arnopo reviewed (commented):
I started reviewing up to the  commit "virtio-msg: add number of virtqueues to GET_DEVICE_INFO"
I will not be able to continue before beginning of next week

bertrand-marquis commented on transport-msg.tex:
```diff
@@ -79,21 +79,21 @@ \subsubsection{System Topology}
   \item \textbf{Bus Instances and Devices}: Each bus instance can advertise
         different capabilities (e.g., transport revision, maximum message size)
         and may discover devices via a message-based or external mechanism.
-        Every discovered device is assigned a \emph{device ID}.
```
Maybe add "uniq": is assigned a uniq device number

wmamills replied:
I am also changing the text to say:
"Every discovered device has a unique \emph{device number}"
Assigned makes it sound like the driver side decides what number to assign.  This is not (normally) the case as the number will come from the device side or DT etc

bertrand-marquis commented on transport-msg.tex:
```diff
@@ -153,40 +153,58 @@ \subsubsection{Transport Revisions and Maximum Message Size}
-        again if necessary.
+        count, the \msgref{SET_CONFIG} request \emph{MUST} be rejected by the
+        device.
+  \item The driver will likely have received \msgref{EVENT_CONFIG} messages
```
Receiving an event in between a request and its response is something that will only occur in some cases and even though it might be hard for the transport to handle it.
It is a possibility but I would remove the "likely" here and do some rephrasing like:
The driver could receive a msg even config and if not it should re-read the configuration count via get_config.

wmamills replied:
Agreed that the race is unlikely.  I think you understand my description of events.  I will try and reword.

What I was trying to describe:
If the device is following all the recommended SHOULDs (not all in the spec yet) then the EVENT_CONFIG _will_ arrive before the response and the driver can just use the new values and figure out if it still want to write.  

The above in NOT the new words for the spec.  I will work on that.

bertrand-marquis commented on transport-msg.tex:
```diff
@@ -317,15 +307,15 @@ \subsubsection{Common Message Format}
     \end{itemize}
   \item \field{msg_id}:
     Uniquely identifies which message definition applies (e.g., GET_DEVICES,
     GET_FEATURES, SET_CONFIG). The specific range or enumeration of types is
```
Here GET_FEATURES must be fixed to GET_DEVICE_FEATURES (this is the case is a number of other places with the change to get/set features) so a global pass should be done on this.
Also the "get devices" being an optional bus message, we might want to gieve an other example like get_device_info

bertrand-marquis commented on transport-msg.tex:
```diff
@@ -350,7 +340,7 @@ \subsubsection{Device Discovery}
   \item By receiving enumeration data from a hypervisor or firmware,
   \item By using \busref{GET_DEVICES} messages (defined in
     \ref{sec:Virtio Transport Options / Virtio Over Messages / Bus Messages})
```
We might want to remove the reference here as \busref is already giving a link.

bertrand-marquis commented on transport-msg.tex:
```diff
@@ -612,29 +602,39 @@ \subsubsection{Driver Notifications}
         information, such as the next descriptor offset or wrap counter
```
There is a feature bit which is configuring if or not those information are available (VIRTIO_F_NOTIFICATION_DATA). Maybe we should mention that those fields are only set if this has been negociated.

wmamills replied:
I did elsewhere.  We can add it again here or reduce the redundancy

wmamills replied:
I added it here again and changed the wording using MUST for feature present and feature absent

bertrand-marquis commented on transport-msg.tex:
```diff
@@ -650,38 +650,41 @@ \subsubsection{Device Notifications}
+The driver \emph{MAY} issue new \msgref{SET_CONFIG}
+messages after initialization if device configuration needs to be
+changed at runtime. The device \emph{MAY} also make configuration changes at
+runtime and \emph{MUST} signal those changes with \msgref{EVENT_CONFIG}
```
I think the "MUST" to signal changes using event_config needs to be lower down to a "SHALL" as we state earlier that there could be cases where this is not done.

wmamills replied:
MUST/SHALL: elsewhere I said MUST send EVENT_CONFIG or EVENT_CONFIG0.  From the Transport level they always get one or the other from the bus.

wmamills replied:
typo missing s??  I don't see it

arnopo replied:
my bad, my brain removed the \emph{...}, my comment is not valid

wmamills replied:
I added EVENT_CONFIG0 here and will probably rename that to EVENT_CONFIG_EMPTY later down.

bertrand-marquis commented on transport-msg.tex:
```diff
@@ -650,38 +650,41 @@ \subsubsection{Device Notifications}
-received and re-initialize the device accordingly.
+an \msgref{EVENT_CONFIG} with a \emph{device status} with the DEVICE_NEEDS_RESET
+bit set.  The driver \emph{SHOULD} reset the device after this using a
+\msgref{SET_DEVICE_STATUS}(0) message.  The driver \emph{MAY} reinitialize the
```
SET_DEVICE_STATUS(0) should be changed to be more coherent to the format used before:
by writing a status of 0 via VIRTIO_MSG_SET_DEVICE_STATUS.

bertrand-marquis commented on transport-msg.tex:
```diff
@@ -737,52 +738,47 @@ \subsubsection{Overview}
 \end{tabular}
 
+Transport message IDs 0x00 to 0x3F are used for messages that require a response
+and IDs 0x40 to 0x7F are used for messages that have no response.  Transport
```
We are using "event" a lot before including the fact that there is not response needed so i would suggest to rephrase here:
0x40 to 0x7F are used for event messages.

bertrand-marquis commented on transport-msg.tex:
```diff
@@ -737,52 +738,47 @@ \subsubsection{Overview}
 \begin{itemize}
-  \item \msgref{ERROR}
   \item \msgref{GET_DEVICE_INFO}
   \item \msgref{GET_FEATURES} and \msgref{SET_FEATURES}
```
needs fixing to GET_DEVICE_FEATURES and SET_DRIVER_FEATURES (other occurrences in the doc)

wmamills replied:
humm, I did do a global search ... but I guess I failed somehow

bertrand-marquis commented on transport-msg.tex:
```diff
@@ -868,16 +875,32 @@ \subsubsection{Overview}
+processing \msgref{EVENT_CONFIG} messages.  If this is not the case it should
+use \msgref{GET_CONFIG} to come up to date.  The driver may then retry the
+\msgref{SET_CONFIG} operation if still desired.
+
```
We are repeating more or less exactly what was explained in the configuration count chapter before.
maybe we should just make a reference to this chapter instead of repeating

arnopo replied:
+1

wmamills replied:
I don't like the duplication but I greatly prefer the details of a message to be with the message.  I would vote we skinny the other section down and keep this one.

wmamills replied:
I did replace this section with a reference.  It is enough here to describe the rejection field value and the message table does that.

bertrand-marquis commented on transport-msg.tex:
```diff
@@ -940,33 +971,56 @@ \subsubsection{Overview}
+\hline
+\end{tabular}
+
+\msgdef{EVENT_CONFIG0}
```
Could we try to find an other name than EVENT_CONFIG0 which is very near to EVENT_CONFIG ?
Maybe event CONFIG_EMPTY and CONFIG_INFO ?

wmamills replied:
EVENT_CONFIG and EVENT_CONFIG0 are the same message with different amounts of data.  I would prefer we just use msg length.  HOWEVER we can use EVENT_CONFIG_EMPTY if you don't like msg length and don't like EVENT_CONFIG0

bertrand-marquis replied:
Using the length is a possibility I do not like because it makes some parts of the message optionals (it is ok when you have a variable length table at the end). And here we must not only remove the data but also the status.
Could we somehow define that a status of 0 means "invalid" (or an other value) so that the transport knows that the status must be retrieved and we would not need CONFIG0 anymore ?

wmamills replied:
Status of 0 is a valid device state so I don't think we should do that.
Our status is 32 bits as is virtio-mmio's field.
The common spec only used values <0xFF so we could define 0x8000_0000 or 0xFFFF_FFFF as invalid but the common spec could invent things for the other bits.

We could add yet another field for valid but I would prefer a different message than that.

I will change the name of EVENT_CONFIG0 to EVENT_CONFIG_EMPTY for now and we can discuss at the sprint.

bertrand-marquis commented on transport-msg.tex:
```diff
@@ -940,33 +971,56 @@ \subsubsection{Overview}
 \msgdef{EVENT_CONFIG}
 
-This message is sent by the virtio-msg backend and does not require a response.
+This message is sent by the virtio-msg device and there is no response.
```
We might need to add something in here to mention that data might be empty (offset = num_bytes = 0)

wmamills replied:
Sure, we can be more explicit.

bertrand-marquis commented on transport-msg.tex:
```diff
@@ -1024,22 +1093,27 @@ \subsection{Bus Messages}\label{sec:Virtio Transport Options / Virtio Over Messa
+for messages with a response and IDs 0xC0 and above are used fro messages with
+no response.
 
 \paragraph{Note:} A bus implementation \textbf{is not required} to use these
```
This is appearing as a numbered chapter in the pdf. We might need to use something else than paragraph here

bertrand-marquis commented on transport-msg.tex:
```diff
@@ -1024,22 +1093,27 @@ \subsection{Bus Messages}\label{sec:Virtio Transport Options / Virtio Over Messa
+
+Bus message IDs 0x80 and above are bus implementation specific. Bus
+implementations \emph{MAY} specify the policy that IDs below 0xC0 be used
+for messages with a response and IDs 0xC0 and above are used fro messages with
```
typo: s/fro/for/

bertrand-marquis commented on transport-msg.tex:
```diff
@@ -1227,24 +1271,20 @@ \subsubsection{Optional Requirements}
+  \item A device or the driver side bus \emph{MUST} send \msgref{EVENT_CONFIG}
+    or \msgref{EVENT_CONFIG0}
+    to inform the driver of configuration of device status changes.
+  \item A device or the driver side bus \emph{MUST} \msgref{EVENT_USED}
```
typo: missing "send" (MUST send EVENT_USED)

bertrand-marquis commented on transport-msg.tex:
```diff
@@ -1269,18 +1309,17 @@ \subsubsection{Compliance for Different Environments}
     (\msgref{GET_DEVICE_INFO}, \msgref{SET_FEATURES}, \msgref{SET_CONFIG}, etc.).
```
SET_FEATURES -> SET_DRIVER_FEATURES

bertrand-marquis reviewed (changes requested):
Very nice, a bunch of minor fixes and suggestions.
We need to make a global pass due to the changes in GET/SET FEATURES as a lot of link and mentions are wrong now.

arnopo commented on transport-msg.tex:
```diff
@@ -449,9 +448,11 @@ \subsubsection{Initialization Flow Overview}
+    determines which features it wants to enable, and writes the desired set
+    with (\msgref{SET_DRIVER_FEATURES}).  It then attempts to set the
+    FEATURES_OK bit in the device status (\msgref{SET_DEVICE_STATUS}) and
+    verifies that it is set in the return device status.
```
I would remove this part . Seems to me that the feature negotiation process is related to the virtio device type not the transport

wmamills replied:
Some of the stuff in our chapter is redundant with then general virtio operation chapter.
Feature negotiation process is NOT device type specific but IS already described in earlier chapters.

I can go either way here.  My college tech writer instructor always told me repetition is NOT bad in tech writing.  (Unlike general prose and code.)

bertrand-marquis replied:
Our transport spec is becoming quite long. I agree with you Bill that repetition is not a bad thing but it has 2 big downsides:
- it must be maintained (and tracking all repetitions when something is changed is complex)
- it must be reviewed (while a reference is easing the reviewer work)
Giving those downside i would definitely vote to remove everything we can that can be replaced by a reference to the same text somewhere in the spec

arnopo commented on transport-msg.tex:
```diff
@@ -216,9 +216,8 @@ \subsubsection{Feature Negotiation Blocks}
-blocks beyond those the device implements, or if the request size exceeds
-the bus instance’s \textbf{maximum message size} limit.
+A device \emph{MUST} respond with all zero feature data for features bit
+requested beyond those the device implements.
```
I would add a sentence to provide driver bus responsability  to respond with 0 if device is not present

wmamills replied:
The driver SHOULD NOT ask for features from a device that does not exist.
The driver CAN ask for DEVICE_INFO from a device that does not exist and we should do better at describing that response: 0 Vendor ID, 0 Device ID, 0 for all sizes.
 I will add this

arnopo commented on transport-msg.tex:
```diff
@@ -495,22 +496,21 @@ \subsubsection{Feature Negotiation}
+set all desired features it tries to set the FEATURES_OK bit in the device status
+via the \msgref{SET_DEVICE_STATUS} message.  The device may accept or reject the
+selected features by returning the new device status with the FEATURES_OK bit
+set (accepted) or cleared (rejected).
```
same here

arnopo commented on transport-msg.tex:
```diff
@@ -932,10 +932,15 @@ \subsubsection{Overview}
+The resulting device status is returned and may not match the status set.  The
+device may set the DEVICE_NEEDS_RESET bit if it has an issue.  The device may
+refuse to set the FEATURES_OK bit if it cannot operate with the features set
+by the driver.
```
regarding
```
2.1.2 Device Requirements: Device Status Field
The device MUST NOT consume buffers or send any used buffer notifications to the driver before DRIVER_-
OK.
The device SHOULD set DEVICE_NEEDS_RESET when it enters an error state that a reset is needed. If
DRIVER_OK is set, after it sets DEVICE_NEEDS_RESET, the device MUST send a device configuration
change notification to the driver.
```
not sure that following sentence will be accepted
```
The device may refuse to set the FEATURES_OK bit if it cannot operate with the features set
by the driver.
```
Here i would just put a reference to chapter `2.1.2`

wmamills replied:
That sentence came directly from some existing part of the spec.  I did not make that up.  I will add a reference to it in our spec here

arnopo replied:
yes `2.1.2 Device Requirements: Device Status Field` is part of the existing doc, 
here we duplicate/change the behavior.  we have to avoid describing a device/driver behavior in the transport layer.

wmamills replied:
We are not changing anything

Example From Section 3.1.1 Driver Requirements: Device Initialization:
5) Set the FEATURES_OK status bit. The driver MUST NOT accept new feature bits after this step.
6) **Re-read device status to ensure the FEATURES_OK bit is still set: otherwise, the device does not support our subset of features and the device is unusable.**

arnopo replied:
Thanks for the reference.  right behavior is similar.  Do you plan to replace the sentence by the reference?

arnopo commented on transport-msg.tex:
```diff
@@ -153,6 +153,13 @@ \subsubsection{Transport Revisions and Maximum Message Size}
+own overhead (e.x. its own header) this is not included in the maximum message
+size. The maximum useful message size is currently expected to be 274.
+This value is large enough to support a GET_CONFIG or SET_CONFIG message with a
+configuration data payload of 256 bytes.
```
regarding

```
2.5.1 Driver Requirements: Device Configuration Space
Drivers MUST NOT assume reads from fields greater than 32 bits wide are atomic, nor are reads from
multiple fields: drivers SHOULD read device configuration space fields like so:
```
Does it makes sense to allow more than one 32-bit word ?

wmamills replied:
Existing transports have a limit of 256 bytes so I think we are OK with 64KB space

wmamills replied:
Our transports allow reading large changes atomically.  Just because other transports can't do that does mean we should stop doing that.

arnopo replied:
Agree on the principle.  Just probably useless regarding the protocol (except if we implement some kind configuration caches). no need to update, let see if there are feedback in upstream on that

arnopo commented on transport-msg.tex:
```diff
@@ -986,6 +992,31 @@ \subsubsection{Overview}
+This message may also be used when a single configuration change on the device
+side cannot be communicated in one \msgref{EVENT_CONFIG} message due to maximum
+message size and does not desire to represent it as multiple configuration
+changes.
```
I read this chapter several time  but this seems to me not obvious for user ( at least to myself)
I would have  one  `VIRTIO_MSG_EVENT_CONFIG` with a generation count field
the device must  wait a `GET_CONFIG` before be able to send another `VIRTIO_MSG_EVENT_CONFIG` 
I would introduce a `VIRTIO_MSG_EVENT_STATUS` with same principle.

arnopo commented on transport-msg.tex:
```diff
@@ -1048,6 +1048,11 @@ \subsubsection{Overview}
+If virtqueue index is 0xFFFF then it indicates that all virtqueus should be
+polled for available buffers. This is useful when the bus implementation is
+using out-of-band notifications, using polling, or when emulating or bridging
+to another virtio transport.
```
this seems breaking the protocol as 0xFFFF is a valid index


```
2.6 Virtqueues
The mechanism for bulk data transport on virtio devices is pretentiously called a virtqueue. Each device can
have zero or more virtqueues1.
A virtio device can have maximum of 65536 virtqueues. Each virtqueue is identified by a virtqueue index. A
virtqueue index has a value in the range of 0 to 65535
```

I would remove this and add
```
Virtqueue index greater than 0xFFFF can be used by the bus for out of band notifications. For instance that  that all virtqueus should be polled for available buffers or when emulating or bridging
to another virtio transport.
```

wmamills replied:
I meant 0xFFFF_FFFF.  Ie -1 but it is not -1 as it is not a signed int.
To be of value here there needs to be an agreed value between bus and transport.

Otherwise the loop can be in the bus and the bus can invent specific virtqueue polls. 
In this system the bus will need to look into the transport and see how many virtqueues it has.

arnopo commented on transport-msg.tex:
```diff
@@ -1053,6 +1053,10 @@ \subsubsection{Overview}
 using out-of-band notifications, using polling, or when emulating or bridging
 to another virtio transport.
 
+The next offset and Next wrap fields should be 0 if VIRTIO_F_NOTIFICATION_DATA
```
s/Next/next/

arnopo reviewed (commented):
Agree with Bertrand it very nice work.
I finish the revue for this set.
As usual my comments are suggestions, that can be discussed /challenged/rejected

wmamills committed:
virtio-msg: review fixes: GET_DEVICE_FEATURES/SET_DEVICE_FEATURES

global search & replace

Signed-off-by: Bill Mills <bill.mills@linaro.org>

wmamills committed:
virtio-msg: review fixes

* add "unique" to device number
* remove "likely" from description of set config race & fix typos
* remove extra reference to bus message section
* remove "implementation" from driver side bus
* mention feature bit that enables extra fields in EVENT_AVAIL
* replace cut & paste "driver" with correct "device"
* mention EVENT_CONFIG0 as other option for notification
* simplify wording around device re-init
* reword SET_CONFIG(0)
* use "event messages" instead of "messages that have no response"
  also use "request/response messages" for the other case
* mention that GET_DEVICE_INFO is the only valid message for an inactive
  device and specify that it should return all zeros
* Remove redundant procedures in SET_CONFIG message defintion and replace
  with reference
* fix spelling of "preferred" and "notify"
* repeat "and the device" to add clarify to EVENT_CONFIG0 def

Signed-off-by: Bill Mills <bill.mills@linaro.org>

wmamills committed:
virtio-msg: rename EVENT_CONFIG0 to EVENT_CONFIG_EMPTY

Reviewers thought the two names were too similar.

Signed-off-by: Bill Mills <bill.mills@linaro.org>

wmamills commented:
On the June 5th meeting it was descided to accept this PR as is and address any remaining fixes in one or more new PRs
